nomad/deploymentwatcher/deployment_watcher.go

```diff
@@ -615,12 +615,30 @@ func (w *deploymentWatcher) handleAllocUpdate(allocs []*structs.AllocListStub) (
 			continue
 		}
 
-		// Determine if the update block for this group is progress based
-		progressBased := dstate.ProgressDeadline != 0
+		hasProgressDeadline := dstate.ProgressDeadline != 0
+		failDeployment := false
+
+		// Fail on the first unhealthy allocation if no progress deadline is specified.
+		if !hasProgressDeadline && alloc.DeploymentStatus.IsUnhealthy() {
+			w.logger.Debug("failing deployment because an allocation failed and the deployment is not progress based", alloc.ID)
+			failDeployment = true
+		}
+
+		if hasProgressDeadline && alloc.DeploymentStatus.IsUnhealthy() &&
+			deployment.Active() {
+			reschedulePolicy := w.j.LookupTaskGroup(alloc.TaskGroup).ReschedulePolicy
+			isRescheduleEligible := alloc.RescheduleEligible(reschedulePolicy, time.Unix(alloc.ModifyTime, 0))
```

Comment:
Can we use `alloc.ModifyTime` here? I'm afraid that this timestamp could also refer to other kind of updates, not only the one we are interested in: the time it turned to Failed 😕

Reply:
Looking at the logic in `node_endpoint.go` and `RescheduleEligible`, shouldn't the time we're using here be the current time? I don't think we care about the time it was last written to Raft (which could be a little bit ago at this point).